src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,221 @@
+            tail = target.prev;
+            tail.next = null;
+        } else {
+            wire(target.prev, target.next);
+        }
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T t) {
-        return false;
+        Node target = getNodeByValue(t);
+        if (target == null) {
+            return false;
+        }
+
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (size == 2) {
+            if (target.value == head.value || target.value.equals(head.value)) {
+                head = tail;
+            } else {
+                tail = head;
+            }
+            wire(head, tail);
+            release(target);
+        } else if (head.value == t || head.value.equals(t)) {
+            head = target.next;
+            head.prev = head;
+            release(target);
+        } else {
+            if (target != tail) {
+                wire(target.prev, target.next);
+            } else if (tail.value == t || tail.value.equals(t)) {
+                tail = target.prev;
+                tail.next = null;
+            }
+            release(target);
+        }
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (size == 0);
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayIndexOutOfBoundsException(index);
+        }
+    }
+
+    private void release(Node node) {
+        node.prev = null;
+        node.next = null;
+    }
+
+    private void wire(Node node1, Node node2) {
+        node1.next = node2;
+        node2.prev = node1;
+    }
+
+    private Node getNodeByIndex(int index) {
+        Node result = null;
+
+        if (index < size / 2) {
+            result = head;
+            for (int i = 0; i <= index; i++) {
+                if (i == index) {
+                    return result;
+                }
+                result = result.next;
+            }
+        } else {
+            result = tail;
+            for (int i = size - 1; i >= index; i--) {
+                if (i == index) {
+                    return result;
+                }
+                result = result.prev;
+            }
+        }
+        return result;
+    }
+
+    private Node getNodeByValue(T value) {
+        Node result = head;
+
+        if (result == null) {
+            return null;
+        }
+        for (int i = 0; i < size; i++) {
+            if (result.value == value
+                    || result.value != null && result.value.equals(value)) {
+                return result;
+            }
+            result = result.next;
+            if (result == null) {
+                return null;
+            }
+        }
+        return result;
+    }
+
+    private class Node {
```

Comment:
```suggestion
    private class Node<T> {
```
let's use parameterized class